File: API/0044_ATR_Range/AtrRangeStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies
{
    /// <summary>
    /// ATR Range strategy
    /// Enters long when price moves up by at least ATR over N candles
    /// Enters short when price moves down by at least ATR over N candles
    /// </summary>
    public class ATRRangeStrategy : Strategy
    {
        private readonly StrategyParam<int> _maPeriod;
        private readonly StrategyParam<int> _atrPeriod;
        private readonly StrategyParam<int> _lookbackPeriod;
        private readonly StrategyParam<DataType> _candleType;
        
        private decimal _nBarsAgoPrice;
        private int _barCounter;

        /// <summary>
        /// MA Period
        /// </summary>
        public int MAPeriod
        {
            get => _maPeriod.Value;
            set => _maPeriod.Value = value;
        }

        /// <summary>
        /// ATR Period
        /// </summary>
        public int ATRPeriod
        {
            get => _atrPeriod.Value;
            set => _atrPeriod.Value = value;
        }

        /// <summary>
        /// Lookback Period (N candles for price movement)
        /// </summary>
        public int LookbackPeriod
        {
            get => _lookbackPeriod.Value;
            set => _lookbackPeriod.Value = value;
        }

        /// <summary>
        /// Candle type for strategy calculation
        /// </summary>
        public DataType CandleType
        {
            get => _candleType.Value;
            set => _candleType.Value = value;
        }

        /// <summary>
        /// Initialize <see cref="ATRRangeStrategy"/>.
        /// </summary>
        public ATRRangeStrategy()
        {
            _maPeriod = Param(nameof(MAPeriod), 20)
                .SetGreaterThanZero()
                .SetDisplay("MA Period", "Period for Moving Average calculation", "Strategy Parameters")
                .SetCanOptimize(true)
                .SetOptimize(10, 50, 10);

            _atrPeriod = Param(nameof(ATRPeriod), 14)
                .SetGreaterThanZero()
                .SetDisplay("ATR Period", "Period for Average True Range calculation", "Strategy Parameters")
                .SetCanOptimize(true)
                .SetOptimize(7, 28, 7);

            _lookbackPeriod = Param(nameof(LookbackPeriod), 5)
                .SetGreaterThanZero()
                .SetDisplay("Lookback Period", "Number of candles to measure price movement", "Strategy Parameters")
                .SetCanOptimize(true)
                .SetOptimize(3, 10, 1);

            _candleType = Param(nameof(CandleType), DataType.TimeFrame(TimeSpan.FromMinutes(5)))
                .SetDisplay("Candle Type", "Type of candles for strategy calculation", "Strategy Parameters");
                
            _nBarsAgoPrice = 0;
            _barCounter = 0;
        }

        /// <inheritdoc />
        public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
        {
            return new[] { (Security, CandleType) };
        }

        /// <inheritdoc />
        protected override void OnStarted(DateTimeOffset time)
        {
            base.OnStarted(time);

            // Create indicators
            var ma = new SimpleMovingAverage { Length = MAPeriod };
            var atr = new AverageTrueRange { Length = ATRPeriod };

            // Create subscription and bind indicators
            var subscription = SubscribeCandles(CandleType);
            
            subscription
                .Bind(ma, atr, ProcessCandle)
                .Start();

            // Configure protection
            StartProtection(
                takeProfit: new Unit(3, UnitTypes.Percent),
                stopLoss: new Unit(2, UnitTypes.Percent)
            );

            // Setup chart visualization
            var area = CreateChartArea();
            if (area != null)
            {
                DrawCandles(area, subscription);
                DrawIndicator(area, ma);
                DrawIndicator(area, atr);
                DrawOwnTrades(area);
            }
        }

        private void ProcessCandle(ICandleMessage candle, decimal maValue, decimal atrValue)
        {
            // Skip unfinished candles
            if (candle.State != CandleStates.Finished)
                return;

            // Check if strategy is ready to trade
            if (!IsFormedAndOnlineAndAllowTrading())
                return;

            // Increment bar counter
            _barCounter++;

            // Store price for first bar of lookback period
            if (_barCounter == 1 || _barCounter % LookbackPeriod == 1)
            {
                _nBarsAgoPrice = candle.ClosePrice;
                LogInfo($"Storing reference price: {_nBarsAgoPrice} at bar {_barCounter}");
                return;
            }

            // Only check for signals at the end of each lookback period
            if (_barCounter % LookbackPeriod != 0)
                return;

            // Calculate price movement over the lookback period
            var priceMovement = candle.ClosePrice - _nBarsAgoPrice;
            var absMovement = Math.Abs(priceMovement);
            
            // Log current values
            LogInfo($"Candle Close: {candle.ClosePrice}, Reference Price: {_nBarsAgoPrice}, Movement: {priceMovement}");
            LogInfo($"ATR: {atrValue}, MA: {maValue}, Absolute Movement: {absMovement}");

            // Check if price movement exceeds ATR
            if (absMovement >= atrValue)
            {
                // Long signal: Price moved up by at least ATR
                if (priceMovement > 0 && Position <= 0)
                {
                    LogInfo($"Buy Signal: Price movement ({priceMovement}) > ATR ({atrValue})");
                    BuyMarket(Volume + Math.Abs(Position));
                }
                // Short signal: Price moved down by at least ATR
                else if (priceMovement < 0 && Position >= 0)
                {
                    LogInfo($"Sell Signal: Price movement ({priceMovement}) < -ATR ({-atrValue})");
                    SellMarket(Volume + Math.Abs(Position));
                }
            }
            
            // Exit logic: Price crosses MA
            if (Position > 0 && candle.ClosePrice < maValue)
            {
                LogInfo($"Exit Long: Price ({candle.ClosePrice}) < MA ({maValue})");
                SellMarket(Math.Abs(Position));
            }
            else if (Position < 0 && candle.ClosePrice > maValue)
            {
                LogInfo($"Exit Short: Price ({candle.ClosePrice}) > MA ({maValue})");
                BuyMarket(Math.Abs(Position));
            }
        }
    }
}

File: API/0044_ATR_Range/atr_range_strategy.py
import clr

clr.AddReference("System.Drawing")
clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan, Math
from System.Drawing import Color
from StockSharp.Messages import UnitTypes, Unit, DataType, ICandleMessage, CandleStates, Sides
from StockSharp.Algo.Indicators import SimpleMovingAverage, AverageTrueRange
from StockSharp.Algo.Strategies import Strategy
from datatype_extensions import *

class atr_range_strategy(Strategy):
    """
    ATR Range strategy
    Enters long when price moves up by at least ATR over N candles
    Enters short when price moves down by at least ATR over N candles
    
    """
    def __init__(self):
        super(atr_range_strategy, self).__init__()
        
        # Initialize internal state
        self._nBarsAgoPrice = 0
        self._barCounter = 0

        # Initialize strategy parameters
        self._maPeriod = self.Param("MAPeriod", 20) \
            .SetDisplay("MA Period", "Period for Moving Average calculation", "Strategy Parameters")

        self._atrPeriod = self.Param("ATRPeriod", 14) \
            .SetDisplay("ATR Period", "Period for Average True Range calculation", "Strategy Parameters")

        self._lookbackPeriod = self.Param("LookbackPeriod", 5) \
            .SetDisplay("Lookback Period", "Number of candles to measure price movement", "Strategy Parameters")

        self._candleType = self.Param("CandleType", tf(5)) \
            .SetDisplay("Candle Type", "Type of candles for strategy calculation", "Strategy Parameters")

    @property
    def MAPeriod(self):
        return self._maPeriod.Value

    @MAPeriod.setter
    def MAPeriod(self, value):
        self._maPeriod.Value = value

    @property
    def ATRPeriod(self):
        return self._atrPeriod.Value

    @ATRPeriod.setter
    def ATRPeriod(self, value):
        self._atrPeriod.Value = value

    @property
    def LookbackPeriod(self):
        return self._lookbackPeriod.Value

    @LookbackPeriod.setter
    def LookbackPeriod(self, value):
        self._lookbackPeriod.Value = value

    @property
    def CandleType(self):
        return self._candleType.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candleType.Value = value

    def OnReseted(self):
        """
        Resets internal state when strategy is reset.
        """
        super(atr_range_strategy, self).OnReseted()
        self._nBarsAgoPrice = 0
        self._barCounter = 0

    def OnStarted(self, time):
        """
        Called when the strategy starts. Sets up indicators, subscriptions, and charting.
        
        :param time: The time when the strategy started.
        """
        super(atr_range_strategy, self).OnStarted(time)

        self._nBarsAgoPrice = 0
        self._barCounter = 0

        # Create indicators
        ma = SimpleMovingAverage()
        ma.Length = self.MAPeriod
        
        atr = AverageTrueRange()
        atr.Length = self.ATRPeriod

        # Create subscription and bind indicators
        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(ma, atr, self.ProcessCandle).Start()

        # Configure protection
        self.StartProtection(
            takeProfit=Unit(3, UnitTypes.Percent),
            stopLoss=Unit(2, UnitTypes.Percent)
        )
        # Setup chart visualization
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, ma)
            self.DrawIndicator(area, atr)
            self.DrawOwnTrades(area)

    def ProcessCandle(self, candle, maValue, atrValue):
        """
        Process candle and execute trading logic
        
        :param candle: The candle message.
        :param maValue: The Moving Average value.
        :param atrValue: The ATR value.
        """
        # Skip unfinished candles
        if candle.State != CandleStates.Finished:
            return

        # Check if strategy is ready to trade
        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        # Increment bar counter
        self._barCounter += 1

        # Store price for first bar of lookback period
        if self._barCounter == 1 or self._barCounter % self.LookbackPeriod == 1:
            self._nBarsAgoPrice = candle.ClosePrice
            self.LogInfo("Storing reference price: {0} at bar {1}".format(self._nBarsAgoPrice, self._barCounter))
            return

        # Only check for signals at the end of each lookback period
        if self._barCounter % self.LookbackPeriod != 0:
            return

        # Calculate price movement over the lookback period
        priceMovement = candle.ClosePrice - self._nBarsAgoPrice
        absMovement = Math.Abs(priceMovement)
        
        # Log current values
        self.LogInfo("Candle Close: {0}, Reference Price: {1}, Movement: {2}".format(
            candle.ClosePrice, self._nBarsAgoPrice, priceMovement))
        self.LogInfo("ATR: {0}, MA: {1}, Absolute Movement: {2}".format(atrValue, maValue, absMovement))

        # Check if price movement exceeds ATR
        if absMovement >= atrValue:
            # Long signal: Price moved up by at least ATR
            if priceMovement > 0 and self.Position <= 0:
                self.LogInfo("Buy Signal: Price movement ({0}) > ATR ({1})".format(priceMovement, atrValue))
                self.BuyMarket(self.Volume + Math.Abs(self.Position))
            # Short signal: Price moved down by at least ATR
            elif priceMovement < 0 and self.Position >= 0:
                self.LogInfo("Sell Signal: Price movement ({0}) < -ATR ({1})".format(priceMovement, -atrValue))
                self.SellMarket(self.Volume + Math.Abs(self.Position))
        
        # Exit logic: Price crosses MA
        if self.Position > 0 and candle.ClosePrice < maValue:
            self.LogInfo("Exit Long: Price ({0}) < MA ({1})".format(candle.ClosePrice, maValue))
            self.SellMarket(Math.Abs(self.Position))
        elif self.Position < 0 and candle.ClosePrice > maValue:
            self.LogInfo("Exit Short: Price ({0}) > MA ({1})".format(candle.ClosePrice, maValue))
            self.BuyMarket(Math.Abs(self.Position))

    def CreateClone(self):
        """
        !! REQUIRED!! Creates a new instance of the strategy.
        """
        return atr_range_strategy()